hmiddleware/logger.go

```diff
@@ -0,0 +1,80 @@
+package hmiddleware
+
+import (
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/go-chi/chi/middleware"
+
+	"github.com/heroku/x/requestid"
+
+	"github.com/sirupsen/logrus"
+)
+
+// PreRequestLogger is a middleware for the github.com/sirupsen/logrus to log requests.
+// It logs things similar to heroku logs and adds remote_addr and user_agent.
+func PreRequestLogger(l logrus.FieldLogger) func(next http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		fn := func(w http.ResponseWriter, r *http.Request) {
+			ww, ok := w.(middleware.WrapResponseWriter)
+			if !ok {
+				ww = middleware.NewWrapResponseWriter(w, r.ProtoMajor)
+			}
+			logRequest(l, r, 0, 0, 0, "start")
+			next.ServeHTTP(ww, r)
+		}
+		return http.HandlerFunc(fn)
+	}
+}
+
+// PostRequestLogger is a middleware for the github.com/sirupsen/logrus to log requests.
+// It logs things similar to heroku logs and adds remote_addr and user_agent.
+func PostRequestLogger(l logrus.FieldLogger) func(next http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		fn := func(w http.ResponseWriter, r *http.Request) {
+			ww, ok := w.(middleware.WrapResponseWriter)
+			if !ok {
+				ww = middleware.NewWrapResponseWriter(w, r.ProtoMajor)
+			}
+
+			t0 := time.Now()
+			defer func() {
+				logRequest(l, r, ww.Status(), ww.BytesWritten(), time.Since(t0), "finish")
+			}()
+			next.ServeHTTP(ww, r)
+		}
+		return http.HandlerFunc(fn)
+	}
+}
+
+func logRequest(l logrus.FieldLogger, r *http.Request, status int, bytes int, service time.Duration, at string) {
+	log := l.WithFields(logrus.Fields{
+		"request_id":  requestid.Get(r),
+		"method":      r.Method,
+		"host":        r.Host,
+		"path":        r.URL.RequestURI(),
```

Comment:
For midgard we try to log the templatized path, so `/apps/{param1}` instead of `/apps/foo`, but in our case we can do that because we are using mux for routing which knows about the path templates. I wonder if there is a generic way to handle that.

Reply:
There's code [in httpmetrics](https://github.com/heroku/x/blob/4b3828443f9722a6c2413023fe79a69c797f33d8/hmiddleware/httpmetrics/httpmetrics.go#L66-L100) that assumes use of chi and extracts route patterns for metrics naming. That could be used there and here to get the pattern/template, if desired.